core/arch/arm/plat-k3/drivers/mailbox.c

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Texas Instruments K3 Mailbox Driver
+ * Copyright (C) 2025 Texas Instruments Incorporated - https://www.ti.com/
+ */
+
+#include <io.h>
+#include <kernel/delay.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "ti_sci_protocol.h"
+#include "ti_sci_transport.h"
+
+static vaddr_t mailbox_tx_base;
+static vaddr_t mailbox_rx_base;
+static vaddr_t mailbox_tx_sram_va;
+static vaddr_t mailbox_rx_sram_va;
+
+static TEE_Result ti_mailbox_poll_rx_status(void)
+{
+	uint32_t num_messages_pending = 0U;
+	vaddr_t mailbox_status_addr = mailbox_rx_base + TI_MAILBOX_MSG_STATUS;
+
+	if (IO_READ32_POLL_TIMEOUT(mailbox_status_addr, num_messages_pending,
+				   num_messages_pending != 0, 10, 1000)) {
+		EMSG("Mailbox RX status polling timed out");
+		return TEE_ERROR_TIMEOUT;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result ti_sci_mailbox_clear_thread(void)
+{
+	uint32_t try_count = 0;
+
+	try_count = io_read32(mailbox_rx_base + TI_MAILBOX_MSG_STATUS);
+	while (io_read32(mailbox_rx_base + TI_MAILBOX_MSG_STATUS) != 0U) {
+		io_read32(mailbox_rx_base + TI_MAILBOX_MSG);
+		if (try_count == 0U) {
+			EMSG("Mailbox RX status polling timed out");
+			return TEE_ERROR_TIMEOUT;
+		}
+		try_count--;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result ti_sci_transport_send(const struct ti_sci_msg *msg)
+{
+	uint32_t num_bytes = 0;
+	paddr_t phys_addr = 0;
+	TEE_Result ret = TEE_SUCCESS;
+
+	FMSG("Clearing the thread");
+	ret = ti_sci_mailbox_clear_thread();
+	if (ret) {
+		EMSG("Thread verification failed. ret = %d", ret);
+		return ret;
+	}
+
+	if (!msg)
```

Comment:
Move this to the top, we shouldn't do anything in this function if we are given bad arguments.